api/src/main/kotlin/org/jetbrains/kotlinx/dl/api/summary/FormatUtils.kt

```diff
@@ -0,0 +1,261 @@
+package org.jetbrains.kotlinx.dl.api.summary
+
+import kotlin.math.max
+
+/**
+ * Convenient function to calculate the required width of the column to fit all the values.
+ * @param [headerWidth] width of the column header
+ * @param [rows] list of rows to calculate the width
+ */
+public fun calcColumnWidth(headerWidth: Int, rows: List<String>): Int {
+    return rows.maxOfOrNull { it.length }?.coerceAtLeast(headerWidth) ?: headerWidth
+}
+
+/**
+ * Convenient function to calculate the required width of all columns.
+ * @param [rows] table with multiple rows each with multiple columns
+ * @param [columnNames] list of column names
+ *
+ * Note: each row must have the same number of columns as the [columnNames] list.
+ * Note: each row may consist of multiple lines.
+ */
+public fun calcColumnWidths(rows: List<TableRow>, columnNames: List<String> = emptyList()): MutableList<Int> {
+    if (columnNames.isEmpty()) {
+        return listOf(rows.flatMap { it.lines() }.maxOfOrNull { it.size } ?: 0).toMutableList()
+    }
+
+    require(rows.all { row ->
+        row.lines().all { it.size == columnNames.size }
+    }) { "All lines must have the same number of columns" }
+
+    return columnNames.indices.map { columnIndex ->
+        val column = rows.map { row -> row.lines() }.flatten().map { it[columnIndex] }
+        calcColumnWidth(columnNames[columnIndex].length, column)
+    }.toMutableList()
+}
+
+/**
+ * Format list of strings to a single line with appropriate paddings for each column.
+ * @param [columnSeparator] sequence of symbols to separate columns
+ * @param [columnsWidths] widths of all columns
+ * @param [strs] list of strings to substitute to columns
+ */
+public fun formatLine(columnSeparator: String, columnsWidths: List<Int>, strs: List<String>): String {
+    return columnsWidths
+        .mapIndexed { index, columnWidth -> (strs.getOrNull(index) ?: "").padEnd(columnWidth) }
+        .joinToString(separator = columnSeparator)
+}
+
+
+/**
+ * Pretty print table from multiple sections.
+ * Each section consist of multiple rows, each row may consist of multiple lines.
+ * Sections may or may not have columns.
+ *
+ * @see [Section]
+ */
+public fun formatTable(vararg sections: Section): List<String> {
+    require(sections.isNotEmpty()) { "At least one section must be provided" }
+
+    val result = mutableListOf<String>()
+
+    val sectionsWithColumns = sections.filterIsInstance<SectionWithColumns>()
+
+    for (section in sectionsWithColumns) {
+        section.alignWith(sectionsWithColumns)
+    }
+
+    val simpleSections = sections.filterIsInstance<SimpleSection>()
+
+    for (section in simpleSections) {
+        section.alignWith(sectionsWithColumns)
```

Comment:
Mm, so the formatting method alters sections passed to it? This would be really unexpected to the caller, let's not do this.

Also, passing information via object fields is super-not-evident. Let's calculate the widths here explicitly, save them to the variable and then pass to the formatting method.